File: problem_set_1/palindrome_pairs.py

def isPalindrome(word):
    return word == word[::-1]

def palindromePairs(words):
    result = []
    for pointer1, word1 in enumerate(words):
        for pointer2, word2 in enumerate(words):
            if pointer2 <= pointer1:
                continue
            if isPalindrome(word1+word2) :
                result.append([pointer1, pointer2])
    return result


if __name__ == '__main__':
    test_cases = [
        ["bat", "tab", "cat"],
        ["dog", "cow", "god", "woc"],
        ["cat", "dog", "god", "tcc"],
        ["cat", "dog", "ogd", "act", "dat"]
    ]
    for words in test_cases:
        print(palindromePairs(words))

File: problem_set_1/palindrome_pairs.go
package main

import "fmt"

func isPalindrome(s string) bool {
    reverse := ""
    for i := len(s) - 1; i >= 0; i-- {
        reverse += string(s[i])
    }
    return s == reverse
}

func palindromePairs(words []string) [][]int {
    result := [][]int{}
    for pointer1, word1 := range words {
        for pointer2, word2 := range words {
            if isPalindrome(word1 + word2) {
                result = append(result, []int{pointer1, pointer2})
            }
        }
    }
    return result
}

func main() {
    test_cases := [][]string{
        {"bat", "tab", "cat"},
        {"dog", "cow", "god", "woc"},
        {"cat", "dog", "god", "tcc"},
        {"cat", "dog", "ogd", "act", "dat"},
    }
    for _, inputWords := range test_cases {
        result := palindromePairs(inputWords)
        fmt.Println(result)
    }
}
